nomad/auth/auth.go

```diff
@@ -0,0 +1,483 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: BUSL-1.1
+
+package auth
+
+import (
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	metrics "github.com/armon/go-metrics"
+	"github.com/hashicorp/go-hclog"
+	"github.com/hashicorp/nomad/acl"
+	"github.com/hashicorp/nomad/helper"
+	"github.com/hashicorp/nomad/nomad/state"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// aclCacheSize is the number of ACL objects to keep cached. ACLs have a parsing
+// and construction cost, so we keep the hot objects cached to reduce the ACL
+// token resolution time.
+const aclCacheSize = 512
+
+type StateGetter func() *state.StateStore
+type LeaderACLGetter func() string
+
+type RPCContext interface {
+	IsTLS() bool
+	IsStatic() bool
+	Certificate() *x509.Certificate
+	GetRemoteIP() (net.IP, error)
+}
+
+type Encrypter interface {
+	VerifyClaim(string) (*structs.IdentityClaims, error)
+}
+
+type Authenticator struct {
+	aclsEnabled  bool
+	tlsEnabled   bool
+	logger       hclog.Logger
+	getState     StateGetter
+	getLeaderACL LeaderACLGetter
+	region       string
+
+	// aclCache is used to maintain the parsed ACL objects
+	aclCache *structs.ACLCache[*acl.ACL]
+
+	// encrypter is a pointer to the server's Encrypter that can be used to
+	// verify claims
+	encrypter Encrypter
+}
+
+func NewAuthenticator(stateFn StateGetter, logger hclog.Logger, getLeaderACLFn LeaderACLGetter, aclsEnabled, tlsEnabled bool, region string, encrypter Encrypter,
+) *Authenticator {
+	return &Authenticator{
+		aclsEnabled:  aclsEnabled,
+		tlsEnabled:   tlsEnabled,
+		logger:       logger.With("auth"),
+		getState:     stateFn,
+		getLeaderACL: getLeaderACLFn,
+		region:       region,
+		aclCache:     structs.NewACLCache[*acl.ACL](aclCacheSize),
+		encrypter:    encrypter,
+	}
+}
+
```

Comment:
Note for reviewers: the code from here down is essentially untouched from the existing `nomad/acl.go`. The only changes should be those that are needed to inject the dependencies from the `nomad` package.